🐛 BUG: `onlineornot checks` should tell you if you're not logged in

### What version of `OnlineOrNot` are you using?

latest

### What operating system are you using?

linux

### Describe the Bug

```bash
rozenmd@thinkpad-pro:~$ npx onlineornot@latest checks

✘ [ERROR] A request to the OnlineOrNot API (/checks) failed.


If you think this is a bug then please create an issue at https://github.com/onlineornot/onlineornot/issues/new
```

```bash
rozenmd@thinkpad-pro:~$ npx onlineornot@latest whoami
 ✅ onlineornot 0.0.10 
----------------------
Getting User settings...
You are not authenticated. Rerun this command with ONLINEORNOT_API_TOKEN set as an environment variable.
```

--- a/.changeset/flat-vans-invite.md
+++ b/.changeset/flat-vans-invite.md
@@ -1,0 +1,5 @@
+---
+"onlineornot": patch
+---
+
+fix: check API token before running commands

--- a/.changeset/flat-vans-invite.md
+++ b/.changeset/flat-vans-invite.md
@@ -3,3 +3,5 @@
 ---
 
 fix: check API token before running commands
+
+This PR makes the error more intuitive when you run commands that call the API without a valid API token.

--- a/packages/onlineornot/src/checks/index.ts
+++ b/packages/onlineornot/src/checks/index.ts
@@ -1,6 +1,7 @@
 import { printBanner } from "../banner";
 import { fetchPagedResult } from "../fetch";
 import { logger } from "../logger";
+import { verifyToken } from "../user";
 import type { Check } from "./types";
 import type {
 	CommonYargsArgv,
@@ -17,6 +18,7 @@ export function checksOptions(yargs: CommonYargsArgv) {
 export async function checksHandler(
 	args: StrictYargsOptionsToInterface<typeof checksOptions>
 ) {
+	await verifyToken();
 	const results = (await fetchPagedResult("/checks")) as Check[];
 
 	if (args.json) {

--- a/packages/onlineornot/src/checks/individualCheck.ts
+++ b/packages/onlineornot/src/checks/individualCheck.ts
@@ -1,6 +1,7 @@
 import { printBanner } from "../banner";
 import { fetchResult } from "../fetch";
 import { logger } from "../logger";
+import { verifyToken } from "../user";
 import type { Check } from "./types";
 import type {
 	CommonYargsArgv,
@@ -23,6 +24,7 @@ export function checkOptions(yargs: CommonYargsArgv) {
 export async function checkHandler(
 	args: StrictYargsOptionsToInterface<typeof checkOptions>
 ) {
+	await verifyToken();
 	const result = (await fetchResult(`/checks/${args.id}`)) as Check;
 
 	if (args.json) {

--- a/packages/onlineornot/src/user/index.ts
+++ b/packages/onlineornot/src/user/index.ts
@@ -1,9 +1,39 @@
 import { getOnlineOrNotAPITokenFromEnv } from "../environment-variables/misc-variables";
+import { fetchResult } from "../fetch";
+
+export const NOT_LOGGED_IN_MSG =
+	"You are not logged in. Rerun this command with ONLINEORNOT_API_TOKEN set as an environment variable.";
+export const INVALID_TOKEN_MSG =
+	"Your API token is invalid. Rerun this command with a different ONLINEORNOT_API_TOKEN set as an environment variable.";
 
 export function getToken() {
 	const token = getOnlineOrNotAPITokenFromEnv();
 	if (!token) {
-		throw new Error("Not logged in.");
+		throw new Error(NOT_LOGGED_IN_MSG);
 	}
 	return { apiToken: token };
+}
+
+export function getTokenQuietly() {
+	const token = getOnlineOrNotAPITokenFromEnv();
+	if (!token) {
+		return undefined;
+	}
+	return { apiToken: token };
+}
+
+//verifyToken loudly fails if a token is missing or invalid
+//good for CLI commands like checks
+//bad for whoami
+export async function verifyToken() {
+	getToken();
+	// verify the token is valid
+	// if not, return undefined
+	try {
+		const result = (await fetchResult("/tokens/verify")) as { status: string };
+		if (result.status !== "active") throw new Error(INVALID_TOKEN_MSG);
+	} catch (e) {
+		//if this fails, token is invalid or expired
+		throw new Error(INVALID_TOKEN_MSG);
+	}
 }

--- a/packages/onlineornot/src/whoami.ts
+++ b/packages/onlineornot/src/whoami.ts
@@ -1,15 +1,15 @@
 import { fetchResult } from "./fetch";
 import { logger } from "./logger";
-import { getToken } from "./user";
+import { NOT_LOGGED_IN_MSG, INVALID_TOKEN_MSG, getTokenQuietly } from "./user";
 
 export async function whoami() {
 	logger.log("Getting User settings...");
 	const user = await getUserInfo();
 
 	if (user === undefined) {
-		return void logger.log(
-			"You are not authenticated. Rerun this command with ONLINEORNOT_API_TOKEN set as an environment variable."
-		);
+		return void logger.log(NOT_LOGGED_IN_MSG);
+	} else if (user === null) {
+		return void logger.log(INVALID_TOKEN_MSG);
 	}
 
 	logger.log(`👋 You are logged in with an ${user.authType}.`);
@@ -29,9 +29,10 @@ export interface UserInfo {
 }
 
 export async function getUserInfo() {
-	const apiToken = getToken();
+	const apiToken = getTokenQuietly();
+	if (!apiToken) return undefined;
 	// verify the token is valid
-	// if not, return undefined
+	// if not, return null
 	let permissions: string[] = [];
 	try {
 		const [verifyResult, permsResult] = await Promise.all([
@@ -43,7 +44,7 @@ export async function getUserInfo() {
 		permissions = permsResult.permissions;
 	} catch (e) {
 		//if this fails, token is invalid or expired
-		return;
+		return null;
 	}
 	return {
 		apiToken: apiToken.apiToken,

--- a/packages/onlineornot/src/whoami.ts
+++ b/packages/onlineornot/src/whoami.ts
@@ -40,7 +40,7 @@ export async function getUserInfo() {
 			fetchResult("/tokens/permissions") as Promise<{ permissions: string[] }>,
 		]);
 
-		if (verifyResult.status !== "active") return;
+		if (verifyResult.status !== "active") return null;
 		permissions = permsResult.permissions;
 	} catch (e) {
 		//if this fails, token is invalid or expired

